File: envoyctl/atomic_write.py
import os
import tempfile


def atomic_write(path, content):
    """Write content to path via a temp file + os.replace().

    os.replace() is an atomic rename on POSIX filesystems, so Envoy's
    inotify-based watch never observes a missing or partially-written file -
    unlike a plain unlink()+move() (delete then create), which has a window
    where the file doesn't exist and can trigger a reload against nothing
    or a truncated read.
    """
    directory = os.path.dirname(os.path.abspath(path))
    fd, tmp_path = tempfile.mkstemp(dir=directory, prefix=".tmp-")
    try:
        with os.fdopen(fd, "w") as f:
            f.write(content)
        os.replace(tmp_path, path)
    except Exception:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)
        raise

File: gatewayctl/internal/atomicwrite/atomicwrite.go
// Package atomicwrite writes a file via a temp file + rename, so a reader
// (Envoy's inotify watch, or config_server serving it) never observes a
// missing or partially-written file mid-swap.
package atomicwrite

import (
	"os"
	"path/filepath"
)

func Write(path string, data []byte) error {
	dir := filepath.Dir(path)
	tmp, err := os.CreateTemp(dir, ".tmp-")
	if err != nil {
		return err
	}
	tmpPath := tmp.Name()

	if _, err := tmp.Write(data); err != nil {
		tmp.Close()
		os.Remove(tmpPath)
		return err
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpPath)
		return err
	}
	return os.Rename(tmpPath, path)
}
